src/sstruct_ls/ssamg_interp.c

```diff
@@ -264,239 +261,138 @@ hypre_SSAMGSetupInterpOp( hypre_SStructMatrix  *A,
 
-         if (P_stencil_size > 1)
+         /* If there are no off-diagonal entries, assemble the matrix and continue outer loop */
+         if (hypre_StructStencilSize(P_stencil) <= 1)
          {
-            if (hypre_StructMatrixConstEntry(P_s, 1))
-            {
-               /* Off-diagonal entries are constant */
-               Pp1 = hypre_StructMatrixConstData(P_s, 1);
-               Pp2 = hypre_StructMatrixConstData(P_s, 2);
+            /* Assemble structured component of prolongation matrix */
+            hypre_StructMatrixAssemble(P_s);
 
-               Pp1[0] = 0.5;
-               Pp2[0] = 0.5;
-            }
-            else
-            {
-               /* Off-diagonal entries are variable */
-               compute_box = hypre_BoxCreate(ndim);
-
-               Pstenc1 = hypre_IndexD(P_stencil_shape[1], cdir);
-               Pstenc2 = hypre_IndexD(P_stencil_shape[2], cdir);
-
-               /* Compute the constant part of the stencil collapse */
-               ventries = hypre_TAlloc(HYPRE_Int, A_stencil_size, HYPRE_MEMORY_HOST);
-               nventries = 0;
-               Pconst[0] = 0.0;
-               Pconst[1] = 0.0;
-               Pconst[2] = 0.0;
-               for (si = 0; si < A_stencil_size; si++)
-               {
-                  if (hypre_StructMatrixConstEntry(A_s, si))
-                  {
-                     Ap = hypre_StructMatrixConstData(A_s, si);
-                     Astenc = hypre_IndexD(A_stencil_shape[si], cdir);
+            /* The following call is needed to prevent cases where interpolation reaches
+            * outside the boundary with nonzero coefficient */
+            hypre_StructMatrixClearBoundary(P_s);
 
-                     if (Astenc == 0)
-                     {
-                        Pconst[0] += Ap[0];
-                     }
-                     else if (Astenc == Pstenc1)
-                     {
-                        Pconst[1] -= Ap[0];
-                     }
-                     else if (Astenc == Pstenc2)
-                     {
-                        Pconst[2] -= Ap[0];
-                     }
-                  }
-                  else
-                  {
-                     ventries[nventries++] = si;
-                  }
-               }
-
-               /* Get stencil space on base grid for entry 1 of P (valid also for entry 2) */
-               hypre_StructMatrixGetStencilSpace(P_s, 1, 0, origin, stride);
+            continue;
+         }
 
-               hypre_CopyToIndex(stride, ndim, Astride);
-               hypre_StructMatrixMapDataStride(A_s, Astride);
-               hypre_CopyToIndex(stride, ndim, Pstride);
-               hypre_StructMatrixMapDataStride(P_s, Pstride);
+         if (hypre_StructMatrixConstEntry(P_s, 1))
+         {
+            /* Off-diagonal entries are constant */
+            hypre_TMemcpy(hypre_StructMatrixConstData(P_s, 1), &half, HYPRE_Complex, 2,
```

Comment:
Is this correct? `half` is just a single value, but this is trying to copy a memory block of size 2?

Reply:
Thanks for catching!